rustls/src/client/ech.rs

```diff
@@ -0,0 +1,509 @@
+use crate::log::trace;
+use crate::msgs::base::{Payload, PayloadU16};
+use crate::msgs::codec::Codec;
+use crate::msgs::enums::{EchVersion, ExtensionType, HpkeAead, HpkeKdf, HpkeKem};
+use crate::msgs::handshake::{
+    ClientExtension, ClientHelloPayload, EchConfig as EchConfigMsg, EchConfigContents,
+    EncryptedClientHello, EncryptedClientHelloOuter, HandshakeMessagePayload, HandshakePayload,
+    HelloRetryRequest, HpkeKeyConfig, HpkeSymmetricCipherSuite, Purpose, Random,
+    ServerHelloPayload, SessionId,
+};
+use crate::msgs::message::{Message, MessagePayload};
+use crate::tls13::key_schedule::{server_ech_hrr_confirmation_secret, KeyScheduleHandshakeStart};
+use crate::{
+    AlertDescription, CommonState, Error, HandshakeType, PeerIncompatible, PeerMisbehaved,
+    ProtocolVersion, Tls13CipherSuite,
+};
+
+/// An enum representing ECH offer status.
+pub(crate) enum Status {
+    /// ECH was not offered - it is a normal TLS handshake.
+    NotOffered,
+    /// ECH was offered and the server accepted.
+    Accepted,
+    /// ECH was offered and the server rejected.
+    Rejected,
+}
+
+/// Randomly generated X25519 public key to use for GREASE ECH HPKE.
+const DUMMY_25519_PUBKEY: &[u8] = &[
+    0x67, 0x35, 0xCA, 0x50, 0x21, 0xFC, 0x4F, 0xE6, 0x29, 0x3B, 0x31, 0x2C, 0xB5, 0xE0, 0x97, 0xD8,
+    0xD0, 0x58, 0x97, 0xCF, 0x5C, 0x15, 0x12, 0x79, 0x4B, 0xEF, 0x1D, 0x98, 0x52, 0x74, 0xDC, 0x5E,
+];
+
+/// Contextual data for a TLS client handshake that has offered encrypted client hello (ECH).
+pub(crate) struct EchState {
+    // The public DNS name from the ECH configuration we've chosen - this is included as the SNI
+    // value for the "outer" client hello. It can only be a DnsName, not an IP address.
+    pub(crate) outer_name: DnsName<'static>,
+    // An HPKE sender context that can be used for encrypting ECH data.
+    sender: Box<dyn HpkeSender>,
+    // The ID of the ECH configuration we've chosen - this is included in the outer ECH extension.
+    config_id: u8,
+    // The private server name we'll use for the inner protected hello.
+    inner_name: ServerName<'static>,
+    // The advertised maximum name length from the ECH configuration we've chosen - this is used
+    // for padding calculations.
+    maximum_name_length: u8,
+    // A supported symmetric cipher suite from the ECH configuration we've chosen - this is
+    // included in the outer ECH extension.
+    cipher_suite: HpkeSymmetricCipherSuite,
+    // A secret encapsulated to the public key of the remote server. This is included in the
+    // outer ECH extension for non-retry outer hello messages.
+    enc: EncapsulatedSecret,
+    // A random value we use for the inner hello.
+    inner_hello_random: Random,
+    // A transcript buffer maintained for the inner hello. Once ECH is confirmed we switch to
+    // using this transcript for the handshake.
+    inner_hello_transcript: HandshakeHashBuffer,
+}
+
+impl EchState {
+    pub(crate) fn new(
```

Comment:
Style nit: reconsider method order? Seems some simpler methods could drift towards the bottom.

Reply:
I reordered slightly, but I think for the most part this is in the preferred order already.